Userland/Utilities/tail.cpp

```diff
@@ -77,19 +52,62 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     TRY(Core::System::pledge("stdio rpath"));
 
     bool follow = false;
-    int line_count = DEFAULT_LINE_COUNT;
-    char const* file = nullptr;
+    size_t line_count = DEFAULT_LINE_COUNT;
+    StringView file;
 
     Core::ArgsParser args_parser;
     args_parser.set_general_help("Print the end ('tail') of a file.");
     args_parser.add_option(follow, "Output data as it is written to the file", "follow", 'f');
     args_parser.add_option(line_count, "Fetch the specified number of lines", "lines", 'n', "number");
-    args_parser.add_positional_argument(file, "File path", "file");
+    args_parser.add_positional_argument(file, "File path", "file", Core::ArgsParser::Required::No);
     args_parser.parse(arguments);
 
-    auto f = TRY(Core::File::open(file, Core::OpenMode::ReadOnly));
-    TRY(Core::System::pledge("stdio"));
+    auto f = TRY(Core::Stream::File::open_file_or_standard_stream(file, Core::Stream::OpenMode::Read));
+    if (!follow)
+        TRY(Core::System::pledge("stdio"));
 
-    auto pos = find_seek_pos(*f, line_count);
-    return tail_from_pos(*f, pos, follow);
+    // Since we can't seek the standard input, we read
+    // the entire file and split on new lines.
+    if (f->is_standard_input()) {
+        do {
+            // FIXME: f->read_all() does not block anymore after
+            // sending EOF (^D), despite f->is_open() returning true.
+            auto buffer = TRY(f->read_all(PAGE_SIZE));
+            auto lines = StringView(buffer).split_view("\n"sv, SplitBehavior::KeepEmpty | SplitBehavior::KeepTrailingSeparator);
+            long i = 0;
+            for (auto line : lines) {
+                if (i >= static_cast<long>(lines.size() - line_count))
+                    out(line);
+                i++;
+            }
+            // Since we can't have FileWatchers on the standard input either,
+            // we just loop forever if the -f option was passed.
+        } while (follow);
+        return 0;
+    }
+
+    auto pos = TRY(find_seek_pos(*f, line_count));
+    dbgln("{}", pos);
+    TRY(tail_from_pos(*f, pos));
+
+    auto seek_to_end = [](Core::Stream::File& file) {
+        MUST(file.seek(0, Core::Stream::SeekMode::FromEndPosition));
+    };
+
+    if (follow) {
+        seek_to_end(*f);
+
+        Core::EventLoop event_loop;
+        auto watcher = TRY(Core::FileWatcher::create());
+        watcher->on_change = [&](Core::FileWatcherEvent const& event) {
+            if (event.type == Core::FileWatcherEvent::Type::ContentModified) {
+                out(MUST(f->read_all()));
+                seek_to_end(*f);
```

Comment:
If this fails we should probably `event_loop->exit()` with a nice exit code instead of hard-crashing the program.